File: Practice/AOJ/ALDS1/ALDS1_12/ALDS1_12A/alds1_12a_kruscal.rb
# aojをpythonでやろうとしたらダメだったので..
# ↑ >0じゃなくて>=0じゃった

Edge = Struct.new(:s, :t, :cost)  # pythonよりrubyのが優れてるところこれ
n = gets.chomp.to_i
edges = []
n.times do |i|
  line = gets.chomp.split.map(&:to_i)
  n.times do |j|
    if line[j] >= 0
      edges << Edge.new(i, j, line[j])
    end
  end
end
edges = edges.sort_by{|e| e.cost}
@parents = (0...n).to_a
def root(x)
  if x == @parents[x]
    return x
  else
    @parents[x] = root(@parents[x])
    return @parents[x]
  end
end
def unite(a, b)
  a, b = root(a), root(b)
  if a!=b
    @parents[b] = a
  end
end
ans = 0
edges.each do |e|
  u, v = e.s, e.t
  if root(u) != root(v)
    unite(u, v)
    ans+=e.cost
  end
end
puts ans

File: Practice/AOJ/ALDS1/ALDS1_12/ALDS1_12A/alds1_12a_kruscal.py
# primよりもクラスカル法のほうが楽らしい
# unionfindを使って追加する辺を判断
# 辺はソートして小さいやつからとっていく
# edgesをコストの昇順でやればwhileはしないでいける？


class Edge:
    def __init__(self, u, v, cost):
        self.u = u
        self.v = v
        self.cost = cost

    def __repr__(self):
        return repr((self.u, self.v, self.cost))


graph = []
n = int(input())
for i in range(n):
    line = list(map(int, input().split()))
    for j in range(n):
        if line[j] >= 0:
            e = Edge(i, j, line[j])
            graph.append(e)

graph = sorted(graph, key=lambda g: g.cost)
parent = [i for i in range(n)]


def root(x):
    if x == parent[x]:
        return parent[x]
    else:
        parent[x] = root(parent[x])
        return parent[x]


def unite(a, b):
    a = root(a)
    b = root(b)
    if a != b:
        parent[b] = a


ans = 0
for g in graph:
    u, v = root(g.u), root(g.v)
    if u != v:
        unite(u, v)
        ans += g.cost

print(ans)
